src/core/net/dnssd_server.cpp

```diff
@@ -0,0 +1,651 @@
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the DNS-SD server.
+ */
+
+#include "dnssd_server.hpp"
+
+#if OPENTHREAD_CONFIG_DNSSD_SERVER_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/debug.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "common/logging.hpp"
+#include "net/srp_server.hpp"
+#include "net/udp6.hpp"
+
+using ot::Encoding::BigEndian::HostSwap16;
+
+namespace ot {
+namespace Dns {
+namespace ServiceDiscovery {
+
+const char Server::kDnssdProtocolUdp[4] = {'_', 'u', 'd', 'p'};
+const char Server::kDnssdProtocolTcp[4] = {'_', 't', 'c', 'p'};
+const char Server::kDefaultDomainName[] = "default.service.arpa.";
+
+Server::Server(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mSocket(aInstance)
+{
+}
+
+otError Server::Start(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(!IsRunning());
+
+    SuccessOrExit(error = mSocket.Open(&Server::HandleUdpReceive, this));
+    SuccessOrExit(error = mSocket.Bind(kPort));
+
+exit:
+    otLogInfoDns("[server] started: %s", otThreadErrorToString(error));
+    return error;
+}
+
+void Server::Stop(void)
+{
+    IgnoreError(mSocket.Close());
+    otLogInfoDns("[server] stopped");
+}
+
+void Server::HandleUdpReceive(void *aContext, otMessage *aMessage, const otMessageInfo *aMessageInfo)
+{
+    static_cast<Server *>(aContext)->HandleUdpReceive(*static_cast<Message *>(aMessage),
+                                                      *static_cast<const Ip6::MessageInfo *>(aMessageInfo));
+}
+
+void Server::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    otError  error = OT_ERROR_NONE;
+    Header   requestHeader;
+    Message *responseMessage = nullptr;
+
+    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), requestHeader));
+    VerifyOrExit(requestHeader.GetType() == Header::kTypeQuery, error = OT_ERROR_DROP);
+
+    responseMessage = mSocket.NewMessage(0);
+    VerifyOrExit(responseMessage != nullptr, error = OT_ERROR_NO_BUFS);
+
+    // Allocate space for DNS header and move offset after DNS header
```

Comment:
nit: where does the "move offset after DNS header" occur?

Reply:
I used to move the offset, but the behavior was removed later.


Comment updated.